datafusion/core/src/physical_planner.rs

```diff
@@ -2006,6 +2001,45 @@ fn tuple_err<T, R>(value: (Result<T>, Result<R>)) -> Result<(T, R)> {
     }
 }
 
+#[derive(Default)]
+struct InvariantChecker;
+
+impl InvariantChecker {
+    /// Checks that the plan change is permitted, returning an Error if not.
+    ///
+    /// In debug mode, this recursively walks the entire physical plan and
+    /// performs additional checks using [`ExecutionPlan::check_node_invariants`].
+    pub fn check(
+        &mut self,
+        plan: &Arc<dyn ExecutionPlan>,
+        rule: &Arc<dyn PhysicalOptimizerRule + Send + Sync>,
```

Comment:
I think for API design we should not pass the rule to the invariant checker (as the checker shouldn't logically depend on the rule). Perhaps just the rule name could be passed in to help with debug messages

Reply:
I expanded the use of `rule` in the invariant check. [See here](https://github.com/apache/datafusion/pull/13986#discussion_r1914279888).